Fix Registrar not removing files when clean_removed is configured

urso opened the pull request:
Closes: #7690, #9215, #10606

This PR includes some changes that also did help to clean/stabilize the `text_clean_removed*` tests. Often times the tests have been stopped to early, due to races between the input file states, registrar states, and test code.

After stabilizing the tests, I reliably (still needs a few runs) can trigger a data-synchronisation issue between the prospectors state and the registry. This is clearly a bug and results in state not being removed from the registry file at all.

After some more testing (yay, heisenbug :/) I found a race condition between harvester shutdown, prospector state cleaning (e.g. if clean_removed is set) and registry updates. The sequence of events goes like this:
```
1. harvester closes file
2. harvester updates prospectors local file states (marks file as finished)
3. prospector starts GC, cleaning old states
4. prospector sends 'remove state event' to the registrar (TTL=0, Finished=True). Note: state is not removed by prospector until next scan completes.
5. harvestar sends 'finished state event' to the registrar (TTL=-1, Finished=True)
6. registrar reads applies state updates:
  1. change state to TTL=0, Finished=True
  2. change state to TTL=-1, Finished=True
  3. clean local state -> file is not removed from registry, due to TTL=-1
  4. write registry
```

The change proposed changes the order the harvester cleanup propagates updates by sending the 'finished event' first before updating the prospectors state. This way we postpone the file being cleaned by the prospector and guarantee the 'finished event' and 'remove event' order (The prospector can not remove the state yet, because Finished is not set until after the 'finished event' has been published).


Additions:
- Add Registry class (accessed via `self.registry()`). This class is used to check/read the registry
- Add InputLogs class for creating/apppending/removing log files used to drive the tests. This reduces path manipulations and reduces repetitive `with open(...) as f:` blocks.
- Add LogState class. This class keeps an offset to start reading from. The offset can be advanced via `checkpoint` or is automatically advanced if `next` succeeds. Using `LogState` we can wait for a particular message to appear in the log, by ignoring old messages. This removes the need to have a count on actual messages.
- Thanks to `LogState` and some minor adjustments to timings I did shave of like 10s from test_registrar.py on my machine.

FB success count:
- test_clean_removed: 5
- test_clean_removed_with_clean_inactive: 5

ruflin commented on filebeat/input/log/harvester.go:
```diff
@@ -391,12 +391,12 @@ func (h *Harvester) SendStateUpdate() {
 	h.publishState(d)
+
+	logp.Debug("harvester", "Update state: %s, offset: %v", h.state.Source, h.state.Offset)
+	h.states.Update(h.state)
```
Can you elaborate on this move? I would assume the harvester itself should be the first one to know about state updates?

urso replied:
Will add my findings to the PR description.

The Update here is no to the harvester, but to the prospector. Thanks to shutdown racing with scan phase we sometimes send finish and remove events out of order if `publishState` is not called before `states.Update`.

urso commented:
Windows is messing with me. The file can not be opened (access denied), which leads to the prospector never cleaning up it's state:

```
2019-02-14T17:17:51.799Z	DEBUG	[input]	log/input.go:477	Start harvester for new file: C:\Users\jenkins\workspace\elastic+beats+pull-request+multijob-windows\beat\filebeat\label\windows\src\github.com\elastic\beats\filebeat\build\system-tests\run\test_registrar.Test.test_clean_removed\log\input1
...
2019-02-14T17:17:52.011Z	ERROR	log/input.go:484	Harvester could not be started on new file: C:\Users\jenkins\workspace\elastic+beats+pull-request+multijob-windows\beat\filebeat\label\windows\src\github.com\elastic\beats\filebeat\build\system-tests\run\test_registrar.Test.test_clean_removed\log\input1, Err: error setting up harvester: Harvester setup failed. Unexpected file opening error: Failed opening C:\Users\jenkins\workspace\elastic+beats+pull-request+multijob-windows\beat\filebeat\label\windows\src\github.com\elastic\beats\filebeat\build\system-tests\run\test_registrar.Test.test_clean_removed\log\input1: Error creating file 'C:\Users\jenkins\workspace\elastic+beats+pull-request+multijob-windows\beat\filebeat\label\windows\src\github.com\elastic\beats\filebeat\build\system-tests\run\test_registrar.Test.test_clean_removed\log\input1': Access is denied.
...
2019-02-14T17:17:52.012Z	DEBUG	[input]	log/input.go:226	State for file not removed because harvester not finished: C:\Users\jenkins\workspace\elastic+beats+pull-request+multijob-windows\beat\filebeat\label\windows\src\github.com\elastic\beats\filebeat\build\system-tests\run\test_registrar.Test.test_clean_removed\log\input1
...
```

urso commented:
Investigating the windows case:
Filebeat used to call Stat in order to check if a file has been removed. This is implemented via `GetFileAttributesEx`. The call always succeeds because the file is still 'visible' in the directory hierarchy + plus the call only checks for the file sizes.
When a file is deleted by another process the 'DeletePending' flag is set. The file will be unlinked and deleted only after all processes have closed the file. This can unfortunately impact log rotation, as no new file with the same name can be generated, as long as the file is still open.

I updated detection of file removal on windows by calling [GetFileInformationByHandleEx](https://docs.microsoft.com/en-us/windows/desktop/api/winbase/nf-winbase-getfileinformationbyhandleex). This call allows us to read the `DeletePending` flag. With this change `close_remove` will ensure the file is closed if the `DeletePending` flag is set. This still requires the reader to actively check for file removal, which is not immediate and might be postponed for an undefinite amount of time, if the outputs are blocking.

This change only slightly improves the validation of `clean_removed`, if at all. Alternatively to this change we can warn/error if the user has set `close_removed` on windows, and skip any `close_removed` related system tests on windows.

Note: I wonder if we can make close_removed more 'stable' by creating a hardlink on windows. But this might not work if network shares or different volumes are used.

ruflin commented on libbeat/common/file/file_windows.go:
```diff
@@ -107,3 +121,28 @@ func ReadOpen(path string) (*os.File, error) {
 	return os.NewFile(uintptr(handle), path), nil
 }
+
+// FileRemoved checks wheter the file held by f is removed.
```
This could need a bit more comments I would say on why `GetFileInformationByHandleEx` needs to be used here. Basically what you describe in the PR.

urso commented:
> Still trying to fully understand the impact. `close_, clean_removed` were not working in the cases where the file was still opened by an other service, if that was not the case, it worked as intended?

On linux a file gets 'unlinked'. This means a file is not visible in the file system anymore, but the inode and data is cleared after the last process has open the file handle. After unlinking a new file with the very same name can be written to.

On windows a file is not 'unlinked', but the 'Delete Pending' flag is set. The file continues to exist within the file system (can still be seen via Glob) as long as at least one process has an open file handle. The 'Delete Pending' flag is not accessible/accounted for when listing a directory or calling `Stat`. That is, filebeat did keep the file open and thusly prevented itself from noticing the file has actually been removed. Worse, as long as filebeat has the file open, no other process can access or replace the file anymore for as long as the 'Delete Pending' flag is set. Here filebeat has the chance to block other processes from making progress (if  file names are to be reused). This we can't really fix.
Due to this `close_removed` setting was ineffective on Windows. The `clean_removed` setting is only effective if filebeat (or no other process) has an open file handle.

After stabilizing the tests, the `test_clean_removed*` tests were failing the very same way on windows, always.

Good news is: filebeat windows tests have been green after fixing `close_removed`. Will restart jenkins a few time and see if I can tune settings a little. I still suspect us having some flakyness due to timing and exponential backoff in the file reader.

urso commented:
Jenkins, test this.

urso commented:
> Could you add a changelog entry?

Still in progress. It has been a rabid hole so far. Let's see if I find more :)

> I assume this will also need to be backported to 7.0 and 6.6 / 6.7?

+1

urso commented:
Jenkins, test this.

urso commented:
Filebeat tests have been green 5 times in a row on all platforms.
